lib/resque/data_store.rb

```diff
@@ -122,11 +122,45 @@ def push_to_queue(queue,encoded_item)
         end
       end
 
-      # Pop whatever is on queue
-      def pop_from_queue(queue)
-        @redis.lpop(redis_key_for_queue(queue))
+      # MLPOP key [key ..]
+      #
+      # LPOP the first available element by probing each key
+      # in the specified order.
+      #
+      # Moving the scheduling logic to the server reduces
+      # the required round-trips especially for workers with
+      # many queues in their watchlists.
+      #
+      # Returns an array of [key-index, element value] or nil
+      # if no element was found.
+      #
+      # Note that we choose to return the key index instead
+      # of the key itself: Redis::Namespace converts the keys
+      # passed to the script (adding a prefix), and we would
+      # have to reverse map the return values. We avoid that
+      # by returning the index instead.
+      def mlpop(*queues, keys:)
+        @redis.evalsha(mlpop_sha, keys: keys)
+      rescue Redis::CommandError => e
+        # If the script is not found (e.g. when the server
+        # is restarted) reload it.
+        if e.message.include?("NOSCRIPT")
+          mlpop_sha!
+          retry
```

Comment:
Let's add some backoff here, failovers take some time to be completed.

Also a timeout (and/or a fallback to the old scheduling) would be nice to avoid a busy loops in case of an error here.

Reply:
The backoff logic in case of a server restart is the responsibility of the redis client.

As far as `mlpop` is concerned, it gets a `NOSCRIPT` error and tries to reload the script. If there is a connectivity issue (e.g. a server restart) the client should kick in, backoff and possibly retry the command internally.

We could add a hidden `retry` argument that decrements itself to guard against busy loops, by I am sure it's needed in this case.